plugins/inputs/cpufreq/cpufreq.go

```
@@ -0,0 +1,157 @@
+)
+
+const defaultHostSys = "/sys"
+
+type CPUFreq struct {
+	Log             telegraf.Logger `toml:"-"`
+	CPUs            []CPU           `toml:"-"`
+	PathSysfs       string          `toml:"host_sys"`
+	GatherThrottles bool            `toml:"gather_throttles"`
+}
+
+type CPU struct {
+	ID    string
+	Path  string
+	Props map[string]*os.File
+}
+
+var sampleConfig = `
+  ## Path for sysfs filesystem.
+  ## See https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt
+  ## Defaults:
+  # host_sys = "/sys"
+
+  ## Gather CPU throttles per core
+  ## Defaults:
+  # gather_throttles = false
+`
+
+func (g *CPUFreq) SampleConfig() string {
+	return sampleConfig
+}
+
+func (g *CPUFreq) Description() string {
+	return "Collect current CPU's frequencies"
+}
+
+func (g *CPUFreq) Init() error {
+	if os.Getenv("HOST_SYS") != "" {
+		g.PathSysfs = os.Getenv("HOST_SYS")
+	} else {
+		g.PathSysfs = defaultHostSys
+	}
+
+	cpus, err := g.DiscoverCpus()
+	if err != nil {
+		return err
+	}
+	g.CPUs = cpus
+
+	return nil
+}
+
+func (g *CPUFreq) Gather(acc telegraf.Accumulator) error {
+	for _, cpu := range g.CPUs {
+		fields := make(map[string]interface{})
+		tags := map[string]string{"cpu": cpu.ID}
+
+		failed := false
+		for name, fd := range cpu.Props {
+			v, err := readUintFromFile(fd)
+			if err != nil {
+				acc.AddError(err)
+				failed = true
+				break
+			}
+
+			fields[name] = v
+		}
+
+		if !failed {
+			acc.AddFields("cpufreq", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (g *CPUFreq) DiscoverCpus() ([]CPU, error) {
+	var cpus []CPU
+
+	cpuDirs, err := filepath.Glob(path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, dir := range cpuDirs {
+		_, cpuName := filepath.Split(dir)
+		cpuNum := strings.TrimPrefix(cpuName, "cpu")
+
+		if _, err := os.Stat(filepath.Join(dir, "cpufreq")); os.IsNotExist(err) {
+			g.Log.Warn("CPUFreq subsystem not available")
+			continue
```

Is it possible that CPUFreq subsystem is not available only for some CPUs and is available for rest of them?

If not, I would return error instead of logging a warning (fail fast ;)).

---

I don't think this is possible in Linux, but I'll check to make sure.